### PR TITLE
AML-1903 AML-2238 Added authorization pipeline

### PR DESCRIPTION
**Pull request by elliotritchie**

### CODE REVIEW

File: src/SFA.DAS.EmployerApprenticeshipsService.Domain/Models/FeatureToggles/Feature.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace SFA.DAS.EAS.Domain.Models.FeatureToggles
+{
+    public class Feature
+    {
+        public FeatureType FeatureType { get; set; }
+
+        public string Name { get; set; }
+
+        public bool Enabled { get; set; }
```
Comment (elliotritchie) [inline src/SFA.DAS.EmployerApprenticeshipsService.Domain/Models/FeatureToggles/Feature.cs]: Is this property used?

File: src/SFA.DAS.EmployerApprenticeshipsService.Domain/Interfaces/IFeatureService.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Threading.Tasks;
+using SFA.DAS.EAS.Domain.Models.FeatureToggles;
+
+namespace SFA.DAS.EAS.Domain.Interfaces
+{
+    /// <summary>
+    ///     Provides access to feature and associated controller actions.
+    /// </summary>
+    public interface IFeatureService
```
Comment (elliotritchie) [inline src/SFA.DAS.EmployerApprenticeshipsService.Domain/Interfaces/IFeatureService.cs]: Could this interface expose just the `GetFeatureThatAllowsAccessToOperationAsync` method instead and allow the `FeatureAgreementService` & `FeatureWhiteListingService` to make the decisions?

File: src/SFA.DAS.EmployerApprenticeshipsService.Domain/Models/FeatureToggles/OperationContext.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿using SFA.DAS.EAS.Domain.Models.Authorization;
+
+namespace SFA.DAS.EAS.Domain.Models.FeatureToggles
+{
+    public class OperationContext
```
Comment (elliotritchie) [inline src/SFA.DAS.EmployerApprenticeshipsService.Domain/Models/FeatureToggles/OperationContext.cs]: Can this be moved to `IAuthorizationContext`?

File: src/SFA.DAS.EmployerApprenticeshipsService.Infrastructure/Pipeline/Features/OperationAuthorisation.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
+﻿using SFA.DAS.EAS.Domain.Models.FeatureToggles;
+using SFA.DAS.NLog.Logger;
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.EAS.Infrastructure.Pipeline.Features
+{
+    public class OperationAuthorisation : IOperationAuthorisationHandler
```
[the remark this answers is not in the document]
Comment (iangledhill) [reply src/SFA.DAS.EmployerApprenticeshipsService.Infrastructure/Pipeline/Features/OperationAuthorisation.cs]: done.

File: src/SFA.DAS.EmployerApprenticeshipsService.Infrastructure/Services/Features/AccountAgreementService.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+            return latestAgreementId;
+        }
+
+        public Task RemoveFromCacheAsync(long accountId)
+        {
+            return _cache.RemoveFromCache(GetCacheKeyForAccount(accountId));
+        }
+
+        private async Task<int?> FetchLatestAgreeemntNumberFromStoreAsync(long accountId)
```
[the remark this answers is not in the document]
Comment (iangledhill) [reply src/SFA.DAS.EmployerApprenticeshipsService.Infrastructure/Services/Features/AccountAgreementService.cs]: Fixed

File: src/SFA.DAS.EmployerApprenticeshipsService.Infrastructure/Services/Features/FeatureAgreementService.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Threading.Tasks;
+using SFA.DAS.EAS.Domain.Interfaces;
+using SFA.DAS.EAS.Domain.Models.FeatureToggles;
+
+namespace SFA.DAS.EAS.Infrastructure.Services.Features
+{
+    public class FeatureAgreementService : IFeatureAgreementService
```
[the remark this answers is not in the document]
Comment (iangledhill) [reply src/SFA.DAS.EmployerApprenticeshipsService.Infrastructure/Services/Features/FeatureAgreementService.cs]: I think the intention was to separate out the service from the pipeline (to allow re-use elsewhere) and to use an adapter pattern to connect the various services into the pipeline. As per discussion you were satisfied with this explanation so am going to leave as is.

File: src/SFA.DAS.EmployerApprenticeshipsService.Infrastructure/Services/Features/FeatureWhiteListingService.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
+﻿using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using SFA.DAS.EAS.Domain.Interfaces;
+using SFA.DAS.EAS.Domain.Models.FeatureToggles;
+
+namespace SFA.DAS.EAS.Infrastructure.Services.Features
+{
+    public class FeatureWhiteListingService : IFeatureWhiteListingService
```
[the remark this answers is not in the document]
Comment (iangledhill) [reply src/SFA.DAS.EmployerApprenticeshipsService.Infrastructure/Services/Features/FeatureWhiteListingService.cs]: Ditto as with FeatureAgreementService above

File: src/SFA.DAS.EAS.Web/DependencyResolution/DefaultRegistry.cs [hunk 8]
```diff
@@ -1,27 +1,251 @@
...
+        }
+
+        private void RegisterPostCodeAnywhereService()
+        {
+            For<IAddressLookupService>().Use<AddressLookupService>();
+            For<IRestClientFactory>().Use<RestClientFactory>();
+            For<IRestServiceFactory>().Use<RestServiceFactory>();
+        }
+
+        private void RegisterAuthorisationPipeline()
```
[the remark this answers is not in the document]
Comment (iangledhill) [reply src/SFA.DAS.EAS.Web/DependencyResolution/DefaultRegistry.cs]: Done

File: src/SFA.DAS.EmployerApprenticeshipsService.Domain/Models/FeatureToggles/Feature.cs (hunk 1 above)
Comment (iangledhill) [reply src/SFA.DAS.EmployerApprenticeshipsService.Domain/Models/FeatureToggles/Feature.cs]: No, it wasn't but it is now used. A new operation authorisation handler has been added that checks this property before any other checks are made.

File: src/SFA.DAS.EmployerApprenticeshipsService.Domain/Interfaces/IFeatureService.cs (hunk 2 above)
Comment (iangledhill) [reply src/SFA.DAS.EmployerApprenticeshipsService.Domain/Interfaces/IFeatureService.cs]: Done

**Pull request merged**